src/simulation_engine.cpp

```diff
@@ -1,21 +1,92 @@
 #include <iostream>
 #include <vector>
 #include <memory>
+#include <random>
+#include <algorithm>
+#include <iterator>
 
 #include <simulator/simulation_engine.hpp>
 #include <simulator/simulation_map.hpp>
+#include <simulator/bug.hpp>
 
 namespace simulation {
 
-SimulationEngine::SimulationEngine() {}
+const uint16_t ANTS = 10;
+const uint16_t DOODLEBUGS = 3;
 
-void SimulationEngine::simulate_round() {}
+SimulationEngine::SimulationEngine(): map_(std::make_shared<SimulationMap>()){}
 
-void SimulationEngine::populate_map_with_species() {}
+void SimulationEngine::start_simulation(uint16_t n_rounds) {
+  for(;n_rounds && simulate_round();n_rounds--);
+}
 
-void SimulationEngine::on_bug_born(std::shared_ptr<Bug>) {}
+bool SimulationEngine::simulate_round() {
+  uint16_t species_left = 0;
+  std::cout << "--------" << std::endl;
+  std::cout << "[SimulationEngine] New Round" << std::endl;
+    map_->print_map();
   
-void SimulationEngine::on_bug_died(std::shared_ptr<Bug>) {}
+  // Iterates over cells and makes every bug move/breed/eat/starve accordingly.
+  for(auto cell: map_->get_cells()) {
+    auto bug_ref = cell->get_bug();
+    if(bug_ref && !bug_ref->has_played()) {
+      bug_ref->play_round();
+    }
+  }
+
+  // Analyzes results of the rounds.
+  bool only_ants = true;
+  bool only_doo = true;
+  for(auto cell: map_->get_cells()) {
+    if(cell->get_bug()) {
+      species_left++;
+      only_doo &= !cell->get_bug()->is_prey();
+      only_ants &= cell->get_bug()->is_prey();
+      cell->get_bug()->reset_turn();
+    }
+  }
+
+  // Checks whether it makes sense to continue the simulation.
+  if(only_doo) {
+    std::cout << "**Doodlebugs Won**" << std::endl;
+    return false;
+  }
+  if(only_ants) {
+    std::cout << "**Ants Won**" << std::endl;
+    return false;
+  }
+  std::cout << "[SimulationEngine]" << "Species left in map: " << species_left << std::endl;
+  std::cout << "--------" << std::endl;
+  return true;
+}
+
+void SimulationEngine::populate_map_with_species() {
+    std::random_device rd;
+    std::mt19937 g(rd());
+    auto cells = map_->get_cells();
+    std::shuffle(cells.begin(), cells.end(), g);
+    auto cells_it = cells.begin();
+    for(auto i=0; i<ANTS; i++) {
+      (*cells_it)->set_bug(std::make_shared<Bug>(*cells_it, BugDescriptorDatabase::get_ant_descriptor()));
+      (*cells_it)->get_bug()->add_observer(this);
+      cells_it++;
+    }
+    for(auto j=0; j<DOODLEBUGS; j++) {
+      (*cells_it)->set_bug(std::make_shared<Bug>(*cells_it, BugDescriptorDatabase::get_doodlebug_descriptor()));
+      (*cells_it)->get_bug()->add_observer(this);
+      cells_it++;
```

Comment:
what happens if you reach the end of iterator?

Reply:
Beware that the `end` iterator points _past _ the last element. See [here](https://en.cppreference.com/w/cpp/container/vector/end)
So if you reach this line you already got a null pointer exception